common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3801,6 +3799,466 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+    if (aead == nullptr) {
+        return nullptr;
+    }
+    if (publicKeyArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "publicKeyArray == null");
+        return {};
+    }
+
+    jbyte* peer_public_key = env->GetByteArrayElements(publicKeyArray, 0);
+
+    jbyte* infoArrayElement = nullptr;
+    const uint8_t* info = nullptr;
+    size_t infoLen = 0;
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    size_t encapsulatedSharedSecretLen;
+    std::vector<uint8_t> encapsulatedSharedSecret(EVP_HPKE_MAX_ENC_LENGTH);
+
+    bool isSetup = EVP_HPKE_CTX_setup_sender(
+            /* ctx= */ ctx,
+            /* out_enc= */ encapsulatedSharedSecret.data(),
+            /* out_enc_len= */ &encapsulatedSharedSecretLen,
+            /* max_enc= */ encapsulatedSharedSecret.size(),
+            /* kem= */ kem,
+            /* kdf= */ kdf,
+            /* aead= */ aead,
+            /* peer_public_key= */ reinterpret_cast<const uint8_t*>(peer_public_key),
+            /* peer_public_key_len= */ env->GetArrayLength(publicKeyArray),
+            /* info= */ info,
+            /* info_len= */ infoLen);
+
+    env->ReleaseByteArrayElements(publicKeyArray, peer_public_key, JNI_ABORT);
+    if (infoArrayElement != nullptr) {
+        env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+    }
+
+    if (!isSetup) {
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "EVP_HPKE_CTX_setup_sender");
+        return {};
+    }
+
+    jbyteArray encArray = env->NewByteArray(encapsulatedSharedSecretLen);
+    env->SetByteArrayRegion(encArray, 0, encapsulatedSharedSecretLen,
+                            reinterpret_cast<const jbyte*>(encapsulatedSharedSecret.data()));
+
+    ScopedLocalRef<jobjectArray> result(
+            env, env->NewObjectArray(2, conscrypt::jniutil::objectClass, nullptr));
+
+    jobject ctxObject = env->NewObject(conscrypt::jniutil::nativeRefHpkeCtxClass,
+                                       conscrypt::jniutil::nativeRefHpkeCtxClass_constructor,
+                                       reinterpret_cast<jlong>(ctx));
+    env->SetObjectArrayElement(result.get(), 0, ctxObject);
+    env->SetObjectArrayElement(result.get(), 1, encArray);
+
+    return result.release();
+}
+
+static jobjectArray NativeCrypto_EVP_HPKE_CTX_setup_sender_with_seed_for_testing(
+        JNIEnv* env, jclass, jint kemValue, jint kdfValue, jint aeadValue,
+        jbyteArray publicKeyArray, jbyteArray infoArray, jbyteArray seedArray) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    JNI_TRACE("EVP_HPKE_CTX_setup_sender_with_seed_for_testing(%d, %d, %d, %p, %p, %p)", kemValue,
+              kdfValue, aeadValue, publicKeyArray, infoArray, seedArray);
+
+    EVP_HPKE_CTX* ctx = EVP_HPKE_CTX_new();
```

Comment:
Ditto on leak

Reply:
Done. Code updated so ctx is now bssl::UniquePtr<EVP_HPKE_CTX>